src/fmt/writer/termcolor/shim_impl.rs

```diff
@@ -28,17 +44,27 @@ impl BufferWriter {
     }
 
     pub(in crate::fmt::writer) fn print(&self, buf: &Buffer) -> io::Result<()> {
-        // This impl uses the `eprint` and `print` macros
-        // instead of using the streams directly.
-        // This is so their output can be captured by `cargo test`
-        let log = String::from_utf8_lossy(&buf.0);
-
-        match self.target {
-            Target::Stderr => eprint!("{}", log),
-            Target::Stdout => print!("{}", log),
-        }
+        if let TargetType::Pipe = self.target {
+            self.target_pipe
+                .as_ref()
+                .unwrap()
+                .lock()
+                .unwrap()
+                .write_all(&buf.0)
+        } else {
+            // This impl uses the `eprint` and `print` macros
+            // instead of using the streams directly.
+            // This is so their output can be captured by `cargo test`
+            let log = String::from_utf8_lossy(&buf.0);
 
-        Ok(())
+            match self.target {
+                TargetType::Stderr => eprint!("{}", log),
+                TargetType::Stdout => print!("{}", log),
+                TargetType::Pipe => unreachable!(),
+            }
+
+            Ok(())
+        }
```

Comment:
You could merge the two (logic) branches into a single match arm:

````suggestion
pub(in crate::fmt::writer) fn print(&self, buf: &Buffer) -> io::Result<()> {
    // This impl uses the `eprint` and `print` macros
    // instead of using the streams directly.
    // This is so their output can be captured by `cargo test`.
    match self.target {
        // Safety: If the target type is `Pipe`, `target_pipe` will always be non-empty.
        TargetType::Pipe => self
            .target_pipe
            .as_ref()
            .unwrap()
            .lock()
            .unwrap()
            .write_all(&buf.0)?,
        TargetType::Stdout => print!("{}", String::from_utf8_lossy(&buf.0)),
        TargetType::Stderr => eprint!("{}", String::from_utf8_lossy(&buf.0)),
    }

    Ok(())
}
````

As far as I can tell the safety does apply to the PR.

Reply:
Ah fluff, I applied the suggested changes backwards and now I can't apply this one directly. (I'm a bit new to reviews)